eth/db/storage.py

```diff
@@ -0,0 +1,470 @@
+        self._db = db
+        self._account_db = account_db
+        self._address = address
+        self._write_trie = None
+        self._starting_root_hash = None  # type: Hash32
+        self._trie_nodes_batch = None  # type: BatchDB
+
+    def _get_write_trie(self) -> HexaryTrie:
+        if self._trie_nodes_batch is None:
+            self._trie_nodes_batch = BatchDB(self._db, read_through_deletes=True)
+
+        if self._write_trie is None:
+            storage_root = self._account_db.get_storage_root(self._address)
+            batched_db = self._trie_nodes_batch
+            self._write_trie = HexaryTrie(batched_db, root_hash=storage_root, prune=True)
+            if self._starting_root_hash is None:
+                self._starting_root_hash = storage_root
+
+        return self._write_trie
+
+    def _get_read_trie(self) -> HexaryTrie:
+        storage_root = self._account_db.get_storage_root(self._address)
+        if storage_root == BLANK_ROOT_HASH:
+            return HexaryTrie({})
+        elif self._write_trie is not None:
+            return self._write_trie
+        else:
+            if self._starting_root_hash is None:
+                self._starting_root_hash = storage_root
+            # Creating "HexaryTrie" is a pretty light operation, so not a huge cost
+            # to create a new one at every read, but we could
+            # cache the read trie, if this becomes a bottleneck.
+            return HexaryTrie(self._db, root_hash=storage_root)
+
+    def _decode_key(self, key: bytes) -> bytes:
+        padded_slot = pad32(key)
+        return keccak(padded_slot)
+
+    def __getitem__(self, key: bytes) -> bytes:
+        hashed_slot = self._decode_key(key)
+        read_trie = self._get_read_trie()
+        try:
+            return read_trie[hashed_slot]
+        except trie_exceptions.MissingTrieNode as exc:
+            self.logger.warning(exc)
+            self.logger.debug("expected to get slot, but:", exc_info=True)
+            raise exc
+
+    def __setitem__(self, key: bytes, value: bytes) -> None:
+        hashed_slot = self._decode_key(key)
+        write_trie = self._get_write_trie()
+        write_trie[hashed_slot] = value
+
+    def _exists(self, key: bytes) -> bool:
+        # used by BaseDB for __contains__ checks
+        hashed_slot = self._decode_key(key)
+        read_trie = self._get_read_trie()
+        return hashed_slot in read_trie
+
+    def __delitem__(self, key: bytes) -> None:
+        hashed_slot = self._decode_key(key)
+        write_trie = self._get_write_trie()
+        try:
+            del write_trie[hashed_slot]
+        except trie_exceptions.MissingTrieNode as exc:
+            self.logger.warning(exc)
+            self.logger.debug("expected to be able to delete slot, but:", exc_info=True)
+
+    @property
+    def has_changed_root(self) -> bool:
+        return self._write_trie and self._write_trie.root_hash != self._starting_root_hash
```

Comment:
*nitpick*: There's no null pointer exception here but there's the possibility that some refactor some day might expose one. This would be ever so slightly safer, and it would support making this method simpler, if `self._starting_root_hash` became a boolean: `self.has_changed_root`

Reply:
I can make the `None` check explicit, which should make a broken refactor more obvious. I'm not as keen on changing `_starting_root_hash` to be a `bool` since it is possible to change to a different root and change back. In that case I expect `has_changed_root` to be `False`. The most obvious way I see to check that is to keep the original root hash around.